src/Groups/matrices/matrix_manipulation.jl

```diff
@@ -0,0 +1,260 @@
+import AbstractAlgebra: FieldElem, map, Ring
+import Hecke: evaluate, multiplicative_jordan_decomposition, PolyElem, _rational_canonical_form_setup, refine_for_jordan
+
+export
+    block_matrix,
+    complement,
+    conjugate_transpose,
+    diagonal_join,
+    insert_block,
+    insert_block!,
+    isconjugate_gl,
+    ishermitian_matrix,
+    isskewsymmetric_matrix,
+    permutation_matrix,
+    submatrix
+
+
+
+########################################################################
+#
+# Matrix manipulation
+#
+########################################################################
+
+"""
+    submatrix(A::MatElem{T}, i::Int, j::Int, m::Int, n::Int)
+Return the `m x n` submatrix of `A` rooted at `(i,j)`
+"""
+function submatrix(A::MatElem, i::Int, j::Int, nr::Int, nc::Int)
+   return matrix(base_ring(A),nr,nc, [A[s,t] for s in i:nr+i-1 for t in j:nc+j-1])
+end
+
+# exists already in Hecke _copy_matrix_into_matrix
+"""
+    insert_block(A::MatElem, B::MatElem, i,j)
+Return the matrix `A` with the block `B` inserted at the position `(i,j)`.
+"""
+function insert_block(A::MatElem{T}, B::MatElem{T}, i::Int, j::Int) where T <: RingElem
+   C = deepcopy(A)
+   for s in 1:nrows(B)
+   for t in 1:ncols(B)
+      C[i+s-1,j+t-1] = B[s,t]
+   end
+   end
+   return C
+end
+
+"""
+    insert_block!(A::MatElem, B::MatElem, i,j)
+Insert the block `B` in the matrix `A` at the position `(i,j)`.
+"""
+function insert_block!(A::MatElem{T}, B::MatElem{T}, i::Int, j::Int) where T <: RingElem
+   for s in 1:nrows(B)
+   for t in 1:ncols(B)
+      A[i+s-1,j+t-1] = B[s,t]
+   end
+   end
+   return A
+end
+
+# exists already in Hecke cat(V...; dims=(1,2))
+"""
+    diagonal_join(V::AbstractVector{<:MatElem})
+    diagonal_join(V::T...) where T <: MatElem
+Return the diagonal join of the matrices in `V`.
+"""
+function diagonal_join(V::AbstractVector{T}) where T <: MatElem
+   nr = sum([nrows(v) for v in V])
+   nc = sum([ncols(v) for v in V])
+   B = zero_matrix(base_ring(V[1]), nr,nc)
+   pos_i=1
+   pos_j=1
+   for k in 1:length(V)
+      insert_block!(B,V[k],pos_i,pos_j)
+      pos_i += nrows(V[k])
+      pos_j += ncols(V[k])
+   end
+   return B
+end
+
+diagonal_join(V::T...) where T <: MatElem = diagonal_join(collect(V))
+
+#=
+diagonal_join(V::T...) where T <: MatElem = cat(V; dims=(1,2))
+=#
+
+"""
+    block_matrix(m::Int, n::Int, V::AbstractVector{T}) where T <: MatElem
+Return the matrix constructed from the given block matrices, which should be given as a sequence `V` of `m*n` matrices (given in row major order, in other words listed across rows). 
```

Comment:
```suggestion
Return the matrix constructed from the given sequence `V` of `m*n` matrices (given in row major order, in other words listed across rows). 
```
Actually, this is ambiguous: does this mean that `length(V)` should be `m*n`; or does it mean that matrices in `V` should all be `m*n` matrices? I assume it's the former, but that should be stated more clearly.

I also am not sure I understand the last part... Perhaps you can elaborate a bit there, too?

Reply:
I just took it from the Magma manual. But I rewrote it right now.